### PR TITLE
Add Bring Your Own Schema support for Composition

### PR DESCRIPTION
**Pull request by barney-s**
### Change description

Add Bring Your Own Schema support
- Add Facade API
  - kubebuilder create api --group composition --version v1alpha1 --kind Facade
  - Add Facade controller
- `openAPIV3Schema`, `facadeKind` to allow the ability to Bring Your Own Schema.
  - Creates a CRD on the fly and is associates with the Composition
  - In additon to the Bring Your Own CRD which we support today
  - Managed CRD option
- Add tests
  - TestSimpleFacadeByoSchema

### Tests you have done

Existing tests pass.

New test added:

```
❯ go test -timeout 3600s -v -run "^TestSimpleFacadeByoSchema$" --images=gcr.io/allotrope-barni/composition:v0.0.1.alpha,gcr.io/allotrope-barni/manifests-inline:v0.0.1.alpha,gcr.io/allotrope-barni/expander-jinja2:v0.0.1.alpha
=== RUN   TestSimpleFacadeByoSchema
    cluster.go:83: Reserved cluster composition-e2e-0
    scenario.go:245: Applying input
    scenario.go:203: Creating object "Namespace/team-a" on cluster "composition-e2e-0"
    scenario.go:216: Creating object "ClusterRole/composition-facade" on cluster "composition-e2e-0"
    scenario.go:216: Creating object "ClusterRoleBinding/composition-facade" on cluster "composition-e2e-0"
    scenario.go:216: Creating object "Facade/default/projectconfigmap" on cluster "composition-e2e-0"
    scenario.go:216: Creating object "Composition/default/projectconfigmap" on cluster "composition-e2e-0"
    scenario.go:216: Creating object "Context/team-a/context" on cluster "composition-e2e-0"
    scenario.go:253: Verifying input
    scenario.go:254: Checking existence of objects ["ClusterRole/composition-facade" "ClusterRoleBinding/composition-facade" "Facade/default/projectconfigmap" "Composition/default/projectconfigmap" "Namespace/team-a" "Context/team-a/context"] on "composition-e2e-0"
    simple_test.go:286: Checking condition "Error" not present in object "Facade/default/projectconfigmap" on cluster "composition-e2e-0"
    scenario.go:276: Loading manifests for: facade pconfig crd
    scenario.go:280: Verifying manifests spec matches: facade pconfig crd
    scenario.go:281: Matching specs of objects ["CustomResourceDefinition/pconfigs.facade.compositions.google.com"] on "composition-e2e-0"
    scenario.go:234: Loading manifests from: facade cr
    scenario.go:216: Creating object "PConfig/team-a/team-a-config" on cluster "composition-e2e-0"
    simple_test.go:297: Checking condition "Error" not present in object "Plan/team-a/pconfigs-team-a-config" on cluster "composition-e2e-0"
    scenario.go:292: Verifying output
    scenario.go:293: Checking existence of objects ["ConfigMap/team-a/proj-a" "ConfigMap/team-a/common-config"] on "composition-e2e-0"
    scenario.go:308: Cleaning up input
    scenario.go:310: Deleting object "ClusterRole/composition-facade" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "ClusterRoleBinding/composition-facade" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "Facade/default/projectconfigmap" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "Composition/default/projectconfigmap" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "Namespace/team-a" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "Context/team-a/context" on cluster "composition-e2e-0"
    scenario.go:312: Checking absence of objects ["ClusterRole/composition-facade" "ClusterRoleBinding/composition-facade" "Facade/default/projectconfigmap" "Composition/default/projectconfigmap" "Namespace/team-a" "Context/team-a/context"] on "composition-e2e-0"
    scenario.go:331: Cleaning up objects: facade pconfig crd
    scenario.go:333: Deleting object "CustomResourceDefinition/pconfigs.facade.compositions.google.com" on cluster "composition-e2e-0"
    scenario.go:335: Checking absence of objects ["CustomResourceDefinition/pconfigs.facade.compositions.google.com"] on "composition-e2e-0"
    scenario.go:331: Cleaning up objects: facade cr
    scenario.go:333: Deleting object "PConfig/team-a/team-a-config" on cluster "composition-e2e-0"
    scenario.go:333: WARN "PConfig/team-a/team-a-config" not found on cluster "composition-e2e-0": the server could not find the requested resource (delete pconfigs.facade.compositions.google.com team-a-config)
    scenario.go:335: Checking absence of objects ["PConfig/team-a/team-a-config"] on "composition-e2e-0"
    scenario.go:319: Cleaning up output
    scenario.go:321: Deleting object "ConfigMap/team-a/proj-a" on cluster "composition-e2e-0"
    scenario.go:321: WARN "ConfigMap/team-a/proj-a" not found on cluster "composition-e2e-0": configmaps "proj-a" not found
    scenario.go:321: Deleting object "ConfigMap/team-a/common-config" on cluster "composition-e2e-0"
    scenario.go:321: WARN "ConfigMap/team-a/common-config" not found on cluster "composition-e2e-0": configmaps "common-config" not found
    scenario.go:323: Checking absence of objects ["ConfigMap/team-a/proj-a" "ConfigMap/team-a/common-config"] on "composition-e2e-0"
    cluster.go:89: Released cluster composition-e2e-0
--- PASS: TestSimpleFacadeByoSchema (31.32s)
PASS
ok      google.com/composition/tests/testcases  115.280s
```

Existing Tests verified:

```
❯ go test -timeout 3600s -v -run "^TestKCCSampleCloudSQL$" --images=gcr.io/allotrope-barni/composition:v0.0.1.alpha,gcr.io/allotrope-barni/manifests-inline:v0.0.1.alpha,gcr.io/allotrope-barni/expander-jinja2:v0.0.1.alpha --use-cc=true
=== RUN   TestKCCSampleCloudSQL
    cluster.go:83: Reserved cluster composition-e2e-0
    scenario.go:245: Applying input
    scenario.go:216: Creating object "Facade/default/sqlha" on cluster "composition-e2e-0"
    scenario.go:216: WARN object exists. Reusing. facades.composition.google.com "sqlha" already exists
    scenario.go:216: Creating object "Composition/default/sqlha" on cluster "composition-e2e-0"
    scenario.go:216: WARN object exists. Reusing. compositions.composition.google.com "sqlha" already exists
    scenario.go:216: Creating object "ClusterRole/composition-facade" on cluster "composition-e2e-0"
    scenario.go:216: WARN object exists. Reusing. clusterroles.rbac.authorization.k8s.io "composition-facade" already exists
    scenario.go:216: Creating object "ClusterRoleBinding/composition-facade" on cluster "composition-e2e-0"
    scenario.go:216: WARN object exists. Reusing. clusterrolebindings.rbac.authorization.k8s.io "composition-facade" already exists
    scenario.go:253: Verifying input
    scenario.go:254: Checking existence of objects ["Facade/default/sqlha" "Composition/default/sqlha" "ClusterRole/composition-facade" "ClusterRoleBinding/composition-facade"] on "composition-e2e-0"
    scenario.go:222: Applying objects: appteam-object
    scenario.go:216: Creating object "CloudSQL/config-control/collateral" on cluster "composition-e2e-0"
    scenario.go:216: Updating already present "CloudSQL/config-control/collateral" on cluster "composition-e2e-0"
    kcc_samples_test.go:182: Checking condition "Ready" present in object "Plan/config-control/cloudsqls-collateral" on cluster "composition-e2e-0"
    client.go:285: Has condition [Ready, ProcessedAllStages, Processed stages: enable-services, block2, block3]
    scenario.go:264: Verifying objects exist: kcc-objects
    scenario.go:265: Checking existence of objects ["SQLInstance/config-control/collateral-db-main" "SQLInstance/config-control/collateral-db-replica-us-central1" "KMSKeyRing/config-control/kmscryptokeyring-us-central1" "KMSKeyRing/config-control/kmscryptokeyring-us-east1" "KMSCryptoKey/config-control/kmscryptokey-enc-us-central1" "KMSCryptoKey/config-control/kmscryptokey-enc-us-east1" "IAMPolicyMember/config-control/sql-kms-us-east1-policybinding" "IAMPolicyMember/config-control/sql-kms-us-central1-policybinding" "ServiceIdentity/config-control/sqladmin.googleapis.com" "Service/config-control/cloudkms.googleapis.com" "Service/config-control/iam.googleapis.com" "Service/config-control/serviceusage.googleapis.com" "Service/config-control/sqladmin.googleapis.com"] on "composition-e2e-0"
    scenario.go:308: Cleaning up input
    scenario.go:310: Deleting object "Facade/default/sqlha" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "Composition/default/sqlha" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "ClusterRole/composition-facade" on cluster "composition-e2e-0"
    scenario.go:310: Deleting object "ClusterRoleBinding/composition-facade" on cluster "composition-e2e-0"
    scenario.go:312: Checking absence of objects ["Facade/default/sqlha" "Composition/default/sqlha" "ClusterRole/composition-facade" "ClusterRoleBinding/composition-facade"] on "composition-e2e-0"
    scenario.go:331: Cleaning up objects: appteam-object
    scenario.go:333: Deleting object "CloudSQL/config-control/collateral" on cluster "composition-e2e-0"
    scenario.go:335: Checking absence of objects ["CloudSQL/config-control/collateral"] on "composition-e2e-0"
    scenario.go:331: Cleaning up objects: kcc-objects
    scenario.go:333: Deleting object "SQLInstance/config-control/collateral-db-main" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "SQLInstance/config-control/collateral-db-replica-us-central1" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "KMSKeyRing/config-control/kmscryptokeyring-us-central1" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "KMSKeyRing/config-control/kmscryptokeyring-us-east1" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "KMSCryptoKey/config-control/kmscryptokey-enc-us-central1" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "KMSCryptoKey/config-control/kmscryptokey-enc-us-east1" on cluster "composition-e2e-0"
    scenario.go:333: WARN "KMSCryptoKey/config-control/kmscryptokey-enc-us-east1" not found on cluster "composition-e2e-0": kmscryptokeys.kms.cnrm.cloud.google.com "kmscryptokey-enc-us-east1" not found
    scenario.go:333: Deleting object "IAMPolicyMember/config-control/sql-kms-us-east1-policybinding" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "IAMPolicyMember/config-control/sql-kms-us-central1-policybinding" on cluster "composition-e2e-0"
    scenario.go:333: Deleting object "ServiceIdentity/config-control/sqladmin.googleapis.com" on cluster "composition-e2e-0"
    scenario.go:333: WARN "ServiceIdentity/config-control/sqladmin.googleapis.com" not found on cluster "composition-e2e-0": serviceidentities.serviceusage.cnrm.cloud.google.com "sqladmin.googleapis.com" not found
    scenario.go:333: Deleting object "Service/config-control/cloudkms.googleapis.com" on cluster "composition-e2e-0"
    scenario.go:333: WARN "Service/config-control/cloudkms.googleapis.com" not found on cluster "composition-e2e-0": services.serviceusage.cnrm.cloud.google.com "cloudkms.googleapis.com" not found
    scenario.go:333: Deleting object "Service/config-control/iam.googleapis.com" on cluster "composition-e2e-0"
    scenario.go:333: WARN "Service/config-control/iam.googleapis.com" not found on cluster "composition-e2e-0": services.serviceusage.cnrm.cloud.google.com "iam.googleapis.com" not found
    scenario.go:333: Deleting object "Service/config-control/serviceusage.googleapis.com" on cluster "composition-e2e-0"
    scenario.go:333: WARN "Service/config-control/serviceusage.googleapis.com" not found on cluster "composition-e2e-0": services.serviceusage.cnrm.cloud.google.com "serviceusage.googleapis.com" not found
    scenario.go:333: Deleting object "Service/config-control/sqladmin.googleapis.com" on cluster "composition-e2e-0"
    scenario.go:333: WARN "Service/config-control/sqladmin.googleapis.com" not found on cluster "composition-e2e-0": services.serviceusage.cnrm.cloud.google.com "sqladmin.googleapis.com" not found
    scenario.go:335: Checking absence of objects ["SQLInstance/config-control/collateral-db-main" "SQLInstance/config-control/collateral-db-replica-us-central1" "KMSKeyRing/config-control/kmscryptokeyring-us-central1" "KMSKeyRing/config-control/kmscryptokeyring-us-east1" "KMSCryptoKey/config-control/kmscryptokey-enc-us-central1" "KMSCryptoKey/config-control/kmscryptokey-enc-us-east1" "IAMPolicyMember/config-control/sql-kms-us-east1-policybinding" "IAMPolicyMember/config-control/sql-kms-us-central1-policybinding" "ServiceIdentity/config-control/sqladmin.googleapis.com" "Service/config-control/cloudkms.googleapis.com" "Service/config-control/iam.googleapis.com" "Service/config-control/serviceusage.googleapis.com" "Service/config-control/sqladmin.googleapis.com"] on "composition-e2e-0"
    scenario.go:319: Cleaning up output
    scenario.go:323: Checking absence of objects [] on "composition-e2e-0"
    cluster.go:89: Released cluster composition-e2e-0
--- PASS: TestKCCSampleCloudSQL (24.00s)
PASS
ok      google.com/composition/tests/testcases  49.783s
```

### CODE REVIEW

**Comment by barney-s** [comment]
For simplifying user guide.

File: experiments/compositions/composition/api/v1alpha1/composition_types.go [hunk 1]
```diff
@@ -119,15 +120,27 @@ type CompositionSpec struct {
...
 
-	Description   string `json:"description,omitempty" protobuf:"bytes,2,name=description"`
-	InputAPIGroup string `json:"inputAPIGroup,omitempty" protobuf:"bytes,3,name=inputAPIGroup"`
+	Description string `json:"description,omitempty"`
+
+	// Use existing KRM API created offline
+	// TODO (barney-s) rename to FacadeAPIGroup,facadeAPIGroup
+	InputAPIGroup string `json:"inputAPIGroup,omitempty"`
+
+	// Bring your own open API spec
```
Comment (cheftako) [inline experiments/compositions/composition/api/v1alpha1/composition_types.go:129]: Nice functionality but would it make more sense to have this as a command line tool?

Comment (barney-s) [reply experiments/compositions/composition/api/v1alpha1/composition_types.go:129]: Absolutely. Kubebuilder/controller-gen is a simpler one to generate APIs. 

Some think it is useful to do something like Gatekeeper ConstraintTemplates. 
At the moment, i am just accommodating those kinds of users. 
The jury is out.

Comment (barney-s) [reply experiments/compositions/composition/api/v1alpha1/composition_types.go:129]: @cheftako - Should we merge this for adding this capability and later align to a simpler schema (custom one as demoed in the call today) ?

Comment (cheftako) [reply experiments/compositions/composition/api/v1alpha1/composition_types.go:129]: I think we are littering the CompositionSpec. At some level I don't think we should care where the CRD came from. If we want another way of generating a CRD, lets put that in a dedicated object but leave composition as streamlined and simple as possible.

Comment (cheftako) [reply experiments/compositions/composition/api/v1alpha1/composition_types.go:129]: Please separate this functionality into its own CRD. Don't want to conflate the composition lifecycle with the CRD lifecycle.

Comment (barney-s) [reply experiments/compositions/composition/api/v1alpha1/composition_types.go:129]: Using `Facade` for the new API for managing CRDs.

Comment (barney-s) [reply experiments/compositions/composition/api/v1alpha1/composition_types.go]: I have separated the BYO Schema creation functionality into a separate `Facade` CRD.

**Comment by barney-s** [comment]
Fixed #1485 1485

**Pull request merged**